src/bin/ion/commands/beta/generate/generator.rs

```diff
@@ -0,0 +1,466 @@
+        }
+    }
+
+    /// Returns true if its a built in type otherwise returns false
+    pub fn is_built_in_type(name: &str) -> bool {
+        matches!(
+            name,
+            "int" | "i64" | "String" | "bool" | "boolean" | "byte[]" | "Vec<u8>" | "float" | "f64"
+        )
+    }
+
+    /// Represents a tera filter that converts given tera string value to upper camel case
+    /// Returns error if the given value is not a string
+    /// For more information: https://docs.rs/tera/1.19.0/tera/struct.Tera.html#method.register_filter
+    pub fn upper_camel(
+        value: &tera::Value,
+        _map: &HashMap<String, tera::Value>,
+    ) -> Result<tera::Value, tera::Error> {
+        Ok(tera::Value::String(
+            value
+                .as_str()
+                .ok_or(tera::Error::msg("Required string for this filter"))?
+                .to_case(Case::UpperCamel),
+        ))
+    }
+
+    /// Generates code for given Ion Schema
+    pub fn generate_code(
+        language: Language,
+        schema: IslSchema,
+        output: &Path,
+    ) -> CodeGenResult<()> {
+        // this will be used for Rust to create mod.rs which lists all the generates modules
+        let mut modules = vec![];
+        let mut module_context = tera::Context::new();
+        let mut code_gen_context = CodeGenContext::new(language);
+
+        // Register a tera filter that can be used to convert a string to upper camel case
+        code_gen_context
+            .tera
+            .register_filter("upper_camel", Self::upper_camel);
+
+        // Create a module for storing all the generated code in the output directory
+        fs::create_dir(output.join("ion_data_model"))?;
+
+        for isl_type in schema.types() {
+            // Initialize the data model as None for each type
+            // this will be filled with appropriate data model when code is generated for the type
+            code_gen_context.data_model = None;
+            Self::generate_data_model(output, &mut modules, &mut code_gen_context, isl_type)?;
+        }
+
+        if code_gen_context.language == Language::Rust {
+            module_context.insert("modules", &modules);
+            let rendered = code_gen_context
+                .tera
+                .render("rust/mod.templ", &module_context)?;
+            let mut file = File::create(output.join("ion_data_model/mod.rs"))?;
+            file.write_all(rendered.as_bytes())?;
+        }
+
+        Ok(())
+    }
+
+    /// Generates data model based on given ISL type definition
+    fn generate_data_model(
+        output: &Path,
+        modules: &mut Vec<String>,
+        code_gen_context: &mut CodeGenContext,
+        isl_type: &IslType,
+    ) -> CodeGenResult<()> {
+        let mut statements = vec![];
+
+        let data_model_name = match isl_type.name().clone() {
+            None => {
+                format!("AnonymousType{}", code_gen_context.anonymous_type_counter)
```

Comment:
We may eventually need this to either be configurable or put all of the anonymous types in their own module to avoid (unlikely) collisions with user types named `AnonymousTypeN`. We can track this in an issue.

Reply:
Reference PR tour comment: https://github.com/amazon-ion/ion-cli/pull/61#discussion_r1391636489
I think there are some alternatives to consider for this and I would love to hear other thoughts on this:
> 🗺️ Template for Rust modules that declares all the modules generated based on given type definitions.This also includes inline type definitions(anonymous type definition) that doesn't have a name.(e.g. `address: { type: string }`) Naming for these types is based on a counter starting with `AnonymousType1`. I have selected this name to imply that these are type definitions that weren't named in schema file, but I am open to other suggestions for naming these types. Some alternatives,
> 
> - Name these types as `InlinedType*`.
> - Make a code generation constraint as open content in schema that allows naming these types (e.g. address: { $codegen_type_name: `address_string`, type: string } )

Reply:
I will create an issue on these alternatives and we can decide an approach for this in a separate PR.

Reply:
Reference #64